### DISCUSSION TITLE
Can't build WASM: System.Exception: Failed to run emscripten

### DISCUSSION THREAD

**Discussion by trydalch** [unoplatform/uno#19087] [Q&A] [answered]
For some time now our WASM build has been broken. Not sure what caused it, but as we've not been developing for web (yet), we've left it untouched. Until now. 

From the binlog, here's the final message that fails:

```
Running [C:\Users\[Username]\...\...\[Project]\obj\Debug\net8.0-browserwasm\workAot]: cmd /c call "C:\Users\[Username]\.uno\emsdk\emsdk-3.1.34\emsdk\emsdk_env.bat" 2>&1 && C:\Users\[Username]\AppData\Local\Temp\dotnet-runtime-wasm-windows-498e18e-7331dcb60e0-8790049905-Release-simd\tools\ninja.exe  -v

[00:00:00.0286732] The system cannot find the path specified.
```

When I check the `.uno\emsdk\emsdk-3.1.34` folder, it's empty.

**Comment by trydalch** [comment]
Here's my binlog. 
[msbuild-redacted.zip](https://github.com/user-attachments/files/18139256/msbuild-redacted.zip)

**Reply by jeromelaban** [reply]
Thanks for the report. This sometimes happens when the git CLI, and can be fixed by removing the `.uno\emsdk` folder. Sometimes it's caused by python not being available. Try running uno-check to get more details.

If it still fails, try taking deleting the folder again and create a binlog, it will likely tell why emscripten is not being installed.

Note that this kind of issues does not exist anymore when using `net9.0`, as the .NET SDK provides its own embedded emscripten/python SDK.

**Reply by trydalch** [reply]
Thanks for the suggestions and explanation. I tried deleting the folder and rebuilding, but got the same error. I'm attaching a new binlog. 

I also ran uno-check and while it updated the templates, android emulators, and downloaded the .net 9 SDK, it didn't resolve this issue (I didn't change my project to .net 9). 

Another interesting bit is I created another sample Uno App and it builds no problem. I can load it in a browser. Searching the binlog for that project and I can't even see a reference to emsdk. 

[msbild-redacted-rebuild.zip](https://github.com/user-attachments/files/18154757/msbild-redacted-rebuild.zip)

**Reply by jeromelaban** [reply]
Thanks. In the failing command, there are multiple locations:
- `C:\Users\REDACTED__Username\source\repos\Z\OGF\Client.Uno\WereAllRelated.Uno\WereAllRelated.Uno\obj\Debug\net8.0-browserwasm\workAot`
- `C:\Users\REDACTED__Username\.uno\emsdk\emsdk-3.1.34\emsdk\emsdk_env.bat`
- `C:\Users\REDACTED__Username\AppData\Local\Temp\dotnet-runtime-wasm-windows-498e18e-7331dcb60e0-8790049905-Release-simd\tools\ninja.exe`

Do they all exist? If they do, it may be related to the redacted path content. We've seen recent issues where the user folder contains accentuated characters and it causes path resolution issue in parts of .NET.

You can change some of the folders by setting a `WasmShellMonoTempFolder` msbuild property to a simple path and an environment variable `WASMSHELL_WSLEMSDK` to another similar simple path (e.g. `c:\s`) and see if it helps.

**Reply by trydalch** [reply]
2/3 of the paths exist. `C:\Users\REDACTED__Username\.uno\emsdk\emsdk-3.1.34\emsdk\emsdk_env.bat`  does not. The folder exists, but there's no `emsdk_env.bat` file. The folder is empty.

**Reply by jeromelaban** [reply]
Thanks. For what I can tell, in the binlog you sent in https://github.com/unoplatform/uno/discussions/19087#discussioncomment-11584251, there already was a folder named `C:\Users\REDACTED__Username\.uno\emsdk\emsdk-3.1.34`:

![image](https://github.com/user-attachments/assets/0d1d5f42-0c49-494e-b22b-8c4bf3aad3f8)

Could you try deleting that folder and create a binlog immediately after that? To be sure you're not getting VS doing parts of it automatically, use the command line `dotnet build -f net8.0-browserwasm /bl`.

Thanks!

**Reply by trydalch** [reply]
I'm getting other odd errors when trying to build via CLI. I suspect it's because we're referencing some old-style .net framework projects and leveraging .Net Compatibility Mode. So I'm not sure I can do that.  Looks like I'm limited to VS. 

A quick aside: When I ran `uno-check` it prompted me to install the .net9 sdk, which I did. It now seems to be using that to build even though my project file is targeting .net8-*. 

I've deleted the folder and then created a binlog immediately after and I'm still getting the same error. The `.uno/emsdk/emsdk-3.1.34` is empty. 

Here's the latest binlog.

[msbuild-12.20.zip](https://github.com/user-attachments/files/18216945/msbuild-12.20.zip)

**Reply by trydalch** [reply, marked as answer]
I think I've figured it out. I have python installed via pyenv-win, so the python that's in my path is another `.bat` file. This is causing the emsdk setup script to exit prematurely, even though there wasn't an error. 

I changed the `python --version` line to:

```cmd
CALL python --version
```

This caused the emsdk setup script to actually run to completion and my wasm build is now fixed.

**Reply by jeromelaban** [reply]
Thanks for the update. Note that in .NET 9, this specific issue is not present anymore, as the Wasm bootstrapper uses the official .NET toolchain with its own self-contained version of the tooling.
